daemon/daemon.go

```diff
@@ -1189,9 +1191,118 @@ func (daemon *Daemon) setHostConfig(container *Container, hostConfig *runconfig.
 	if err := daemon.RegisterLinks(container, hostConfig); err != nil {
 		return err
 	}
-
+	var (
+		volumes []*vv
+		binds   = make(map[string]struct{})
+	)
+	for _, v := range hostConfig.VolumesFrom {
+		containerID, mode, err := parseVolumesFrom(v)
+		if err != nil {
+			return err
+		}
+		c, err := daemon.Get(containerID)
+		if err != nil {
+			return err
+		}
+		vf, err := container.addVolumesFrom(c, mode)
+		if err != nil {
+			return err
+		}
+		volumes = append(volumes, vf...)
+	}
+	hasVolumeFrom := func(dest string) bool {
+		for _, vf := range volumes {
+			if vf.vc.Destination == dest {
+				return true
+			}
+		}
+		return false
+	}
+	var (
+		uniqueVolumes       []volume.Volume
+		uniqueVolumeConfigs = make(map[string]*VolumeConfig)
+	)
+	for _, v := range container.volumes {
+		config := container.VolumeConfig[v.Name()]
+		if !hasVolumeFrom(config.Destination) {
+			uniqueVolumes = append(uniqueVolumes, v)
+			uniqueVolumeConfigs[v.Name()] = config
+		}
+	}
+	container.volumes = uniqueVolumes
+	container.VolumeConfig = uniqueVolumeConfigs
+	for _, vf := range volumes {
+		if _, exists := container.VolumeConfig[vf.v.Name()]; exists {
+			continue
+		}
+		container.volumes = append(container.volumes, vf.v)
+		container.VolumeConfig[vf.v.Name()] = vf.vc
+	}
+	uniqueVolumes = []volume.Volume{}
```

Comment:
all the logic to handle bind mounts should be wrapper into a big `if hostConfig.Binds() > 0`, from 1241 to 1272. Otherwise, volumes mounted with `volumesFrom` and `volume` can be overridden with an empty array when there are no bind mounts.